FWCore/Utilities/scripts/edmCheckClassTransients

```diff
@@ -0,0 +1,108 @@
+#!  /usr/bin/env python
+import string
+import re
+from optparse import OptionParser
+
+class XmlParser(object):
+    """Parses a classes_def.xml file looking for class declarations that contain
+    ClassVersion attributes. Once found looks for sub-elements named 'version'
+    which contain the ClassVersion to checksum mappings.
+    """
+    
+    #The following are constants used to describe what data is kept
+    # in which index in the 'classes' member data
+    originalNameIndex=0
+    classVersionIndex=1
+    versionsToChecksumIndex = 2
+    
+    def __init__(self,filename):
+        self._file = filename
+        self.classes = dict()
+        self.anames =[]
+        self._presentClass = None
+        self.readClassesDefXML()
+    def readClassesDefXML(self):
+        import xml.parsers.expat
+        p = xml.parsers.expat.ParserCreate()
+        p.StartElementHandler = self.start_element
+        p.EndElementHandler = self.end_element
+        f = open(self._file)
+        # Replace any occurence of <>& in the attribute values by the xml parameter
+        rxml, nxml = f.read(), ''
+        q1,q2 = 0,0
+        for c in rxml :
+            if   (q1 or q2) and c == '<' : nxml += '&lt;'
+            elif (q1 or q2) and c == '>' : nxml += '&gt;'
+            # elif (q1 or q2) and c == '&' : nxml += '&amp;'
+            else                         : nxml += c
+            if c == '"' : q1 = not q1
+            if c == "'" : q2 = not q2
+        try : p.Parse(nxml)
+        except xml.parsers.expat.ExpatError, e :
+            print '--->> edmCheckClassVersion: ERROR: parsing selection file ',self._file
+            print '--->> edmCheckClassVersion: ERROR: Error is:', e
+            raise
+        f.close()
+    def start_element(self,name,attrs):
+        if name in ('class','struct'):
+            if 'name' in attrs:
+                if re.search('^edm::AssociationMap<',attrs['name']):
```

Comment:
OK. I was thinking of making a dictionary of classes and transient fields.


On Sep 21, 2015, at 10:41 AM, Chris Jones <notifications@github.com<mailto:notifications@github.com>> wrote:


In FWCore/Utilities/scripts/edmCheckClassTransients<https://github.com/cms-sw/cmssw/pull/11397#discussion_r39986273>:

> +            if   (q1 or q2) and c == '<' : nxml += '&lt;'
> +            elif (q1 or q2) and c == '>' : nxml += '&gt;'
> +            # elif (q1 or q2) and c == '&' : nxml += '&amp;'
> +            else                         : nxml += c
> +            if c == '"' : q1 = not q1
> +            if c == "'" : q2 = not q2
> +        try : p.Parse(nxml)
> +        except xml.parsers.expat.ExpatError, e :
> +            print '--->> edmCheckClassVersion: ERROR: parsing selection file ',self._file
> +            print '--->> edmCheckClassVersion: ERROR: Error is:', e
> +            raise
> +        f.close()
> +    def start_element(self,name,attrs):
> +        if name in ('class','struct'):
> +            if 'name' in attrs:
> +                if re.search('^edm::AssociationMap<',attrs['name']):


Let's make this more extensible. How about putting the pattern to search into a list and loop over the list? That way as new cases arise we just add to the list.

—
Reply to this email directly or view it on GitHub<https://github.com/cms-sw/cmssw/pull/11397/files#r39986273>.